File: iosApp/iosApp/view/screens/notes/masterEditor/MasterEditorScreen.swift
import SwiftUI
import shared

struct MasterEditorScreen: View {

    let noteId: String?
    var onOpenMedia: (String?) -> Void = { _ in }

    @StateObject private var viewModel = MasterEditorViewModel()
    @Environment(Router.self) var router: Router
    @Environment(\.colorScheme) private var scheme
    @Environment(\.dismiss) private var dismiss
    @State private var showAttach = false
    @State private var sheet: MasterTextSheet = .none

    private var palette: SmartTextPalette { SmartTextPalette.of(scheme) }

    private var commands: CanvasCommands { CanvasCommands.shared }

    var body: some View {
        ZStack(alignment: .bottom) {
            canvas
            banner
            zoomBar
            keyboardToolbar
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(palette.page.ignoresSafeArea())
        .onAppear { viewModel.load(noteId: noteId) }
        .confirmationDialog(
            "Add beside the focused widget",
            isPresented: $showAttach,
            titleVisibility: .visible
        ) {
            attachActions
        }
        .editorCapabilities(
            state: viewModel.capabilities,
            noteTitle: viewModel.note?.title ?? "this note",
            callbacks: EditorCapabilityCallbacks(
                onState: { viewModel.onCapabilityState($0) },
                onOpenCamera: {
                    router.navigate(to: .Camera { media in viewModel.onCaptured(media) })
                },
                onCaptured: { viewModel.onCaptured($0) },
                onFilesPicked: { viewModel.importFiles(urls: $0) },
                onImportLink: { viewModel.importFromLink($0) },
                onDeleteContent: { viewModel.deleteContent($0) },
                onDeleteNote: { dismiss() }
            )
        )
        .sheet(isPresented: Binding(
            get: { sheet.isPresented },
            set: { if !$0 { sheet = .none } }
        )) {
            if let editingId = viewModel.canvas.editingNodeId,
               let focused = viewModel.text(for: editingId) {
                MasterTextSheetHost(
                    sheet: sheet,
                    toolbar: focused.toolbar,
                    onIntent: { viewModel.onTextIntent(nodeId: editingId, intent: $0) },
                    onDismiss: { sheet = .none }
                )
            }
        }
    }

    private var canvas: some View {
        MasterCanvas(
            state: viewModel.canvas,
            onIntent: viewModel.onCanvasIntent,
            onRename: { nodeId, name in viewModel.renameNode(nodeId: nodeId, name: name) }
        ) { node, isEditing in
            nodeBody(node: node, isEditing: isEditing)
        }
    }

    private func nodeBody(node: CanvasNode, isEditing: Bool) -> MasterNodeContent {
        let nodeId: String = node.id
        let scale = CGFloat(viewModel.canvas.viewport.scale)
        return MasterNodeContent(
            node: node,
            isEditing: isEditing,
            scale: scale,
            textState: viewModel.text(for: nodeId),
            content: viewModel.content(for: node),
            dismissToken: viewModel.keyboardDismissToken,
            onTextIntent: { intent in viewModel.onTextIntent(nodeId: nodeId, intent: intent) },
            onFocused: { viewModel.onCanvasIntent(commands.setEditing(nodeId: nodeId)) },
            onOpenMedia: { onOpenMedia(node.contentId) },
            onDelete: { viewModel.deleteNode(nodeId: nodeId) }
        )
    }

    /// Shown only while the canvas has nothing to draw. Tells us which stage of
    /// load -> read note -> read canvas -> measure viewport actually failed.
    @ViewBuilder
    private var banner: some View {
        if viewModel.canvas.document.nodes.isEmpty {
            VStack(alignment: .leading, spacing: 4) {
                Text("No canvas nodes yet")
                    .font(.system(size: 14, weight: .semibold))
                Text("note id: \(viewModel.note?.id ?? "nil")")
                Text("contents: \(viewModel.noteContents.count)  ·  texts: \(viewModel.texts.count)")
                Text("viewport: \(Int(viewModel.canvas.viewport.widthPx))×\(Int(viewModel.canvas.viewport.heightPx)) @ \(viewModel.canvas.zoomPercent)%")
                Text("loading: \(viewModel.isLoading ? "yes" : "no")")
                if let message = viewModel.errorMessage {
                    Text(message).foregroundColor(palette.accent)
                }
            }
            .font(.system(size: 12, design: .monospaced))
            .foregroundColor(palette.onSurface)
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(palette.codeBackground)
            .frame(maxHeight: .infinity, alignment: .top)
        }
    }

    @ViewBuilder
    private var keyboardToolbar: some View {
        if let editingId = viewModel.canvas.editingNodeId,
           let focused = viewModel.text(for: editingId) {
            SmartTextKeyboardToolbar(
                toolbar: focused.toolbar,
                expanded: focused.isToolbarExpanded,
                canUndo: focused.canUndo,
                canRedo: focused.canRedo,
                onAction: { action in handle(action, nodeId: editingId, state: focused) }
            )
        }
    }

    @ViewBuilder
    private var attachActions: some View {
        Button("New page") { viewModel.addPage() }
        Button("Photo or video") { viewModel.requestCapture(ContentType.image) }
        Button("Record audio") { viewModel.requestCapture(ContentType.audio) }
        Button("Location") { viewModel.requestCapture(ContentType.location) }
        Button("Import a file") { viewModel.openImportSheet() }
        Button("Table") { viewModel.addWidget(kind: ContentType.table) }
        Button("Drawing") { viewModel.addWidget(kind: ContentType.drawing) }
        Button("Rebuild layout from note order") { viewModel.rebuildLayoutFromNote() }
        Button("Apply canvas order back to the note") { viewModel.applyCanvasOrderToNote() }
        Button("Cancel", role: .cancel) {}
    }

    private var zoomLabel: String { "\(viewModel.canvas.zoomPercent)%" }

    private var zoomBar: some View {
        HStack(spacing: 4) {
            barButton(icon: "minus", tint: palette.onSurface) {
                viewModel.onCanvasIntent(commands.zoomOut())
            }
            Text(zoomLabel)
                .font(.system(size: 13))
                .foregroundColor(palette.onSurfaceMuted)
            barButton(icon: "plus", tint: palette.onSurface) {
                viewModel.onCanvasIntent(commands.zoomIn())
            }
            barButton(icon: "viewfinder", tint: palette.onSurface) {
                viewModel.onCanvasIntent(commands.zoomToFit())
            }
            toolButtons
            barButton(icon: "plus.square.on.square", tint: palette.accent) {
                showAttach = true
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(RoundedRectangle(cornerRadius: 12).fill(palette.toolbar))
        .padding(.bottom, 28)
    }

    private var toolButtons: some View {
        ForEach(commands.tools(), id: \.self) { tool in
            let active = commands.isTool(state: viewModel.canvas, tool: tool)
            barButton(icon: icon(for: tool), tint: active ? palette.accent : palette.onSurface) {
                viewModel.onCanvasIntent(commands.setTool(tool: tool))
            }
        }
    }

    private func icon(for tool: CanvasTool) -> String {
        switch tool {
        case CanvasTool.hand: return "hand.raised.fill"
        case CanvasTool.zoom: return "plus.magnifyingglass"
        case CanvasTool.lock: return "lock.fill"
        default: return "hand.point.up.left"
        }
    }

    private func barButton(
        icon: String,
        tint: Color,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            Image(systemName: icon).foregroundColor(tint)
        }
    }

    private func handle(_ action: ToolbarAction, nodeId: String, state: MasterTextState) {
        let master = MasterTextCommands.shared
        if SmartTextCommands.shared.isMore(action: action) {
            viewModel.onTextIntent(
                nodeId: nodeId,
                intent: master.setToolbarExpanded(expanded: !state.isToolbarExpanded)
            )
        } else if SmartTextCommands.shared.isDismiss(action: action) {
            viewModel.keyboardDismissToken &+= 1
            viewModel.onCanvasIntent(commands.setEditing(nodeId: nil))
        } else if let intent = master.forToolbar(action: action) {
            viewModel.onTextIntent(nodeId: nodeId, intent: intent)
        } else {
            // style / colour / size / align / link have no intent — they open a sheet
            let opened = MasterTextSheet.of(action: action)
            if opened.isPresented {
                viewModel.keyboardDismissToken &+= 1
                sheet = opened
            }
        }
    }
}

struct MasterNodeContent: View {

    let node: CanvasNode
    let isEditing: Bool
    let scale: CGFloat
    let textState: MasterTextState?
    let content: NoteContentModel?
    let dismissToken: Int
    let onTextIntent: (MasterTextIntent) -> Void
    let onFocused: () -> Void
    var onOpenMedia: () -> Void = {}
    var onDelete: () -> Void = {}

    @Environment(\.colorScheme) private var scheme
    @Environment(\.openURL) private var openURL

    private var palette: SmartTextPalette { SmartTextPalette.of(scheme) }

    private var media: NoteContentModel.MediaContent? {
        content as? NoteContentModel.MediaContent
    }

    @ViewBuilder
    var body: some View {
        if content is NoteContentModel.TextContent {
            textBody
        } else if let media {
            switch media.type {
            case ContentType.image, ContentType.gif, ContentType.video, ContentType.audio:
                mediaBody

            case ContentType.docx, ContentType.epub, ContentType.md,
                 ContentType.pdf, ContentType.other:
                InlineBookFileView(
                    media: media,
                    onOpenFull: onOpenMedia,
                    onDelete: onDelete
                )

            default:
                placeholder("Media")
            }
        } else if content is NoteContentModel.Link {
            linkBody
        } else if content is NoteContentModel.Location {
            locationBody
        } else {
            placeholder(node.kind.name)
        }
    }

    @ViewBuilder
    private var textBody: some View {
        if let textState {
            // NOT readOnly: !isEditing — flipping editability as the view becomes first
            // responder tears down and rebuilds the input session, which closed the
            // keyboard and reopened it. Focus drives editingNodeId, never the reverse.
            MasterTextWidget(
                state: textState,
                readOnly: false,
                scale: scale,
                accessory: nil,
                dismissToken: dismissToken,
                shouldFocus: isEditing,
                onFocused: onFocused,
                onIntent: onTextIntent
            )
            .padding(16)
        } else {
            placeholder("Empty text")
        }
    }

    @ViewBuilder
    private var mediaBody: some View {
        if let media {
            switch media.type {
            case ContentType.image, ContentType.gif:
                CardImageEditor(
                    content: media,
                    actionClick: onOpenMedia,
                    actionDelete: onDelete
                )

            case ContentType.video:
                VideoCardPlayer(
                    content: media,
                    actionClick: onOpenMedia,
                    actionDelete: onDelete
                )

            case ContentType.audio:
                AudioPlayView(mediaContent: media, onDelete: onDelete)

            default:
                placeholder("Media")
            }
        } else {
            placeholder("Media")
        }
    }

    private var linkBody: some View {
        let link = content as? NoteContentModel.Link
        return VStack(alignment: .leading, spacing: 8) {
            Image(systemName: "link").foregroundColor(palette.accent)
            Text(link?.url.isEmpty == false ? link!.url : "Link")
                .font(.system(size: 15))
                .foregroundColor(palette.accent)
                .lineLimit(3)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
        .padding(16)
        .contentShape(Rectangle())
        .onTapGesture {
            if let raw = link?.url, let url = URL(string: raw) { openURL(url) }
        }
    }

    private var locationBody: some View {
        let location = content as? NoteContentModel.Location
        let label: String = {
            if let address = location?.address, !address.isEmpty { return address }
            if let location { return "\(location.latitude), \(location.longitude)" }
            return "Location"
        }()
        return VStack(alignment: .leading, spacing: 8) {
            Image(systemName: "mappin.and.ellipse").foregroundColor(palette.accent)
            Text(label)
                .font(.system(size: 15))
                .foregroundColor(palette.onSurface)
                .lineLimit(3)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
        .padding(16)
        .contentShape(Rectangle())
        .onTapGesture {
            guard let location,
                  let url = URL(string: "maps://?ll=\(location.latitude),\(location.longitude)")
            else { return }
            openURL(url)
        }
    }

    private func placeholder(_ label: String) -> some View {
        ZStack {
            palette.codeBackground
            Text(label)
                .font(.system(size: 13))
                .foregroundColor(palette.onSurfaceMuted)
        }
    }
}

File: androidApp/src/main/java/com/app/pustakam/android/screen/masterEditor/MasterEditorScreen.kt
package com.app.pustakam.android.screen.masterEditor

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.CenterFocusStrong
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.PanTool
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material.icons.filled.TouchApp
import androidx.compose.material.icons.filled.ZoomIn
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material.icons.filled.ZoomOut
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.app.pustakam.android.screen.editor.EditorCapabilityCallbacks
import com.app.pustakam.android.screen.editor.EditorCapabilityHost
import com.app.pustakam.android.screen.editor.permissionsFor
import com.app.pustakam.android.widgets.masterEditor.MasterCanvas
import com.app.pustakam.android.widgets.smartText.SmartTextKeyboardToolbar
import com.app.pustakam.android.widgets.smartText.SmartTextSheet
import com.app.pustakam.android.widgets.smartText.SmartTextSheetHost
import com.app.pustakam.core.richtext.master.presentation.MasterTextCommands
import com.app.pustakam.core.richtext.presentation.SmartTextCommands
import com.app.pustakam.android.widgets.smartText.SmartTextTokens
import com.app.pustakam.core.common.util.ContentType

import com.app.pustakam.core.richtext.master.presentation.CanvasCommands
import com.app.pustakam.core.richtext.master.presentation.CanvasTool

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MasterEditorScreen(
    noteId: String? = null,
    viewModel: MasterEditorViewModel = viewModel(),
    onBack: () -> Unit = {},
    onOpenMedia: (String?) -> Unit = {},
    onCaptureMedia: (String?) -> Unit = {}
) {
    val colors = SmartTextTokens.colors
    val uiState by viewModel.state.collectAsStateWithLifecycle()
    val canvas = uiState.canvas
    var showAttach by remember { mutableStateOf(false) }
    var sheet by remember { mutableStateOf(SmartTextSheet.NONE) }

    LaunchedEffect(noteId) { viewModel.load(noteId) }

    EditorCapabilityHost(
        state = uiState.capabilities,
        noteTitle = uiState.note?.title.orEmpty(),
        permissions = permissionsFor(uiState.capabilities.pendingCapture),
        audioDraft = { uiState.audioDraft },
        callbacks = EditorCapabilityCallbacks(
            onState = viewModel::onCapabilityState,
            onOpenCamera = { onCaptureMedia(uiState.note?.id) },
            onAudioSaved = viewModel::onCaptured,
            onFilesPicked = { context, uris -> viewModel.importDeviceFiles(context, uris) },
            onImportLink = { context, link -> viewModel.importFromLink(context, link) },
            onDeleteContent = viewModel::deleteContent,
            onDeleteNote = { onBack() }
        )
    )

    Box(modifier = Modifier.fillMaxSize().background(colors.page)) {
        MasterCanvas(
            state = canvas,
            onIntent = viewModel::onCanvasIntent,
            onRename = viewModel::renameNode
        ) { node, isEditing ->
            MasterNodeContent(
                node = node,
                isEditing = isEditing,
                scale = canvas.viewport.scale,
                textState = uiState.textFor(node.id),
                content = uiState.note?.contents?.firstOrNull { it.id == node.contentId },
                onTextIntent = { viewModel.onTextIntent(node.id, it) },
                onFocused = { viewModel.onCanvasIntent(CanvasCommands.setEditing(node.id)) },
                onOpenMedia = { onOpenMedia(node.contentId) },
                onDelete = { viewModel.deleteNode(node.id) }
            )
        }

        Row(
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .navigationBarsPadding()
                .padding(horizontal = 12.dp, vertical = 16.dp)
                .background(colors.toolbar, RoundedCornerShape(12.dp))
                .horizontalScroll(rememberScrollState())
                .padding(horizontal = 6.dp, vertical = 2.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(2.dp)
        ) {
            IconButton(onClick = { viewModel.onCanvasIntent(CanvasCommands.zoomOut()) }) {
                Icon(Icons.Default.ZoomOut, contentDescription = "Zoom out", tint = colors.onSurface)
            }
            Text(
                text = "${canvas.zoomPercent}%",
                style = TextStyle(color = colors.onSurfaceMuted, fontSize = 13.sp)
            )
            IconButton(onClick = { viewModel.onCanvasIntent(CanvasCommands.zoomIn()) }) {
                Icon(Icons.Default.ZoomIn, contentDescription = "Zoom in", tint = colors.onSurface)
            }
            IconButton(
                onClick = { showAttach = true },

            ) {
                Icon(Icons.Default.Add, contentDescription = "Zoom in", tint = colors.onSurface)
            }
            IconButton(onClick = { viewModel.onCanvasIntent(CanvasCommands.zoomToFit()) }) {
                Icon(
                    Icons.Default.CenterFocusStrong,
                    contentDescription = "Fit to screen",
                    tint = colors.onSurface
                )
            }
            CanvasCommands.tools().forEach { tool ->
                val active = canvas.tool == tool
                IconButton(onClick = { viewModel.onCanvasIntent(CanvasCommands.setTool(tool)) }) {
                    Icon(
                        imageVector = when (tool) {
                            CanvasTool.SELECT -> Icons.Default.TouchApp
                            CanvasTool.HAND -> Icons.Default.PanTool
                            CanvasTool.ZOOM -> Icons.Default.ZoomIn
                            CanvasTool.LOCK -> Icons.Default.Lock
                        },
                        contentDescription = CanvasCommands.toolLabel(tool),
                        tint = if (active) colors.accent else colors.onSurface
                    )
                }
            }
        }
        val focusedText = canvas.editingNodeId?.let { uiState.textFor(it) }
        if (focusedText != null) {
            SmartTextKeyboardToolbar(
                toolbar = focusedText.toolbar,
                expanded = focusedText.isToolbarExpanded,
                canUndo = focusedText.canUndo,
                canRedo = focusedText.canRedo,
                onAction = { action ->
                    val nodeId = canvas.editingNodeId ?: return@SmartTextKeyboardToolbar
                    if (SmartTextCommands.isMore(action)) {
                        viewModel.onTextIntent(
                            nodeId,
                            MasterTextCommands.setToolbarExpanded(!focusedText.isToolbarExpanded)
                        )
                    } else if (SmartTextCommands.isDismiss(action)) {
                        viewModel.onCanvasIntent(CanvasCommands.setEditing(null))
                    } else {
                        val intent = MasterTextCommands.forToolbar(action)
                        if (intent != null) {
                            viewModel.onTextIntent(nodeId, intent)
                        } else {
                            sheet = SmartTextSheet
                                .fromIndex(MasterTextCommands.sheetIndex(action))
                        }
                    }
                },
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .imePadding()
            )
        }

        val sheetNodeId = canvas.editingNodeId
        if (sheet != SmartTextSheet.NONE && focusedText != null && sheetNodeId != null) {
            SmartTextSheetHost(
                sheet = sheet,
                currentStyle = focusedText.toolbar.paragraphStyle,
                currentAlign = focusedText.toolbar.align,
                currentFontSize = focusedText.toolbar.fontSize,
                currentLink = focusedText.toolbar.link,
                searchQuery = "",
                replacement = "",
                matchCount = 0,
                currentMatch = 0,
                tableRowCount = 0,
                tableColumnCount = 0,
                onDismiss = { sheet = SmartTextSheet.NONE },
                onStyle = {
                    viewModel.onTextIntent(sheetNodeId, MasterTextCommands.setParagraphStyle(it))
                    sheet = SmartTextSheet.NONE
                },
                onAlign = {
                    viewModel.onTextIntent(sheetNodeId, MasterTextCommands.setAlignment(it))
                    sheet = SmartTextSheet.NONE
                },
                onColor = {
                    val intent = if (sheet == SmartTextSheet.BACKGROUND_COLOR) {
                        MasterTextCommands.setBackgroundColor(it)
                    } else {
                        MasterTextCommands.setTextColor(it)
                    }
                    viewModel.onTextIntent(sheetNodeId, intent)
                    sheet = SmartTextSheet.NONE
                },
                onFontSize = {
                    viewModel.onTextIntent(sheetNodeId, MasterTextCommands.setFontSize(it))
                    sheet = SmartTextSheet.NONE
                },
                onLink = {
                    viewModel.onTextIntent(sheetNodeId, MasterTextCommands.setLink(it))
                    sheet = SmartTextSheet.NONE
                },
                onTable = { sheet = SmartTextSheet.NONE },
                onInsertTable = { _, _ -> sheet = SmartTextSheet.NONE },
                onSearchQuery = {},
                onReplacement = {},
                onFindNext = {},
                onFindPrevious = {},
                onReplaceCurrent = {},
                onReplaceAll = {}
            )
        }

        if (showAttach) {
            ModalBottomSheet(onDismissRequest = { showAttach = false }) {
                Column(modifier = Modifier.padding(bottom = 24.dp)) {
                    Text(
                        text = "Add beside the focused widget",
                        style = TextStyle(color = colors.onSurfaceMuted, fontSize = 13.sp),
                        modifier = Modifier.padding(horizontal = 20.dp, vertical = 8.dp)
                    )
                    Text(
                        text = "Rebuild layout from note order",
                        style = TextStyle(color = colors.accent, fontSize = 16.sp),
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable {
                                showAttach = false
                                viewModel.rebuildLayoutFromNote()
                            }
                            .padding(horizontal = 20.dp, vertical = 14.dp)
                    )
                    Text(
                        text = "Apply canvas order back to the note",
                        style = TextStyle(color = colors.accent, fontSize = 16.sp),
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable {
                                showAttach = false
                                viewModel.applyCanvasOrderToNote()
                            }
                            .padding(horizontal = 20.dp, vertical = 14.dp)
                    )
                    listOf(
                        "New page" to ContentType.TEXT,
                        "Table" to ContentType.TABLE,
                        "Drawing" to ContentType.DRAWING
                    ).forEach { (label, kind) ->
                        Text(
                            text = label,
                            style = TextStyle(color = colors.onSurface, fontSize = 16.sp),
                            modifier = Modifier
                                .fillMaxWidth()
                                .clickable {
                                    showAttach = false
                                    viewModel.addWidget(kind)
                                }
                                .padding(horizontal = 20.dp, vertical = 14.dp)
                        )
                    }

                    listOf(
                        "Photo or video" to ContentType.IMAGE,
                        "Record audio" to ContentType.AUDIO,
                        "Location" to ContentType.LOCATION
                    ).forEach { (label, kind) ->
                        Text(
                            text = label,
                            style = TextStyle(color = colors.onSurface, fontSize = 16.sp),
                            modifier = Modifier
                                .fillMaxWidth()
                                .clickable {
                                    showAttach = false
                                    viewModel.requestCapture(kind)
                                }
                                .padding(horizontal = 20.dp, vertical = 14.dp)
                        )
                    }
                }
            }
        }
    }
}
